SIP-28: Background Events (#150)
MetaMask/SIPs: SIPS/sip-28.md

---
sip: 28
title: Background Events
status: Draft
author: Olaf Tomalka (@ritave)
created: 2024-10-09
---

## Abstract

> This SIP introduces a way the schedule non-recurring events in the future.

Reviewer: ```suggestion
This SIP introduces a way to  schedule non-recurring events in the future.
```

## Motivation

> Scheduled recurring events are already support in Snaps through Cronjobs feature. Introducing non-recurring events we allow novel use-cases for Snap developers, such as allowing a snap that sets a reminder for ENS domain expiration date.

Reviewer: ```suggestion
Scheduled recurring events are already supported in the Snaps platform via the Cronjobs feature. By introducing non-recurring events we will allow novel use-cases for Snap developers, such as allowing a snap that sets a reminder for ENS domain expiration date.
```

## Specification

> Indented sections like this are considered non-normative.

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### Snap Manifest

> This SIP introduces a new permission `endowment:background-events`. This permissions grants a Snap the ability to schedule future events.

Reviewer: ```suggestion
This SIP introduces a new permission `endowment:background-events`. This permission grants a Snap the ability to schedule future events.
```

> This SIP introduces a new permission `endowment:background-events`. This permissions grants a Snap the ability to schedule future events.

Reviewer: Wondering if we should somehow merge this with cronjobs 🤔 Otherwise, we'll have two permissions that do extremely similar things 😄

Reviewer: WDYT @Mrtenz?

Reviewer: I think we eventually want to rename cronjobs to background events, but doing so will be a breaking change.

This SIP introduces a new permission `endowment:background-event`. This permissions grants a Snap the ability to schedule future events.

This permission takes no parameters and is specified in the `snap.manifest.json` as follows:

```json
{
  "initialPermissions": {

>     "endowment:background-event": {}

Reviewer 2: Pluralise maybe?
```suggestion
    "endowment:background-events": {}
```

  }
}
```

### RPC Methods

> #### `snap_backgroundEventSchedule`

Reviewer 2: ```suggestion
#### `snap_scheduleBackgroundEvent`
```

> #### `snap_scheduleBackgroundEvent`

Reviewer 2: Please add a TypeScript definition for this.

This method allows a Snap to schedule a callback to `onBackgroundEvent` handler in the future with a JSON-RPC request object as a parameter.

The RPC method takes two parameters:

> - `date` - An [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) date and time and optional timezone offset.

Reviewer: Do we assume UTC or local time if no timezone is provided?

Author: In ISO8601 if no time-zone is provided, it's local time, if there's a `Z` timezone, it means UTC.

`2024-10-09T09:59:38Z` vs `2024-10-09T09:59:38`

Reviewer: Ah, makes sense!

Reviewer 3: I think that defaulting to localtime (which may not always be canonically known) can become problematic and would like to propose that there should be no ambiguity in interpretation:
- #152

> - `date` - An [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) date and time and optional timezone offset.

Reviewer: Is `date` preferred to letting the caller specify a number of seconds or something?

Wondering what is the simplest for developers.

Author: * Notice that we have minute rather than second level precision.
* Conversion to ISO strings is a single call to `date.toISOString()`.
* What happens to seconds when daylight time change happens?
* What happens to seconds when user travels between timezones (local-time vs UTC distinction is an edge case here, where local-time changes while UTC one doesn't)

  - The time's precision SHALL be truncated on the extension side to minutes.
- `request` - A JSON object that will provided as-is to `onBackgroundEvent` handler as parameter.

```typescript
snap.request({
  method: "snap_backgroundEventSchedule",
  params: {
    date: "2024-10-09T09:59",
    request: {
      method: "foobar",
      params: {
        foo: "bar",
      },
    },
  },
});
```

The RPC method call returns a `string` that is a unique ID representing that specific background event, allowing it to be cancelled.

> #### `snap_backgroundEventCancel`

Reviewer 2: ```suggestion
#### `snap_cancelBackgroundEvent`
```

> #### `snap_cancelBackgroundEvent`

Reviewer 2: Please add a TypeScript definition for this.

This method allows to cancel an already scheduled background event using the unique ID returned from `snap_backgroundEventSchedule`

This RPC method takes one argument:

- `id` - The id that was returned during schedule.

```typescript
snap.request({
  method: "snap_backgroundEventCancel",
  params: {
    id: myReturnedId,
  },
});
```

> ### `onBackgroundEvent` handler

Reviewer 2: Please add a TypeScript definition for this.

This SIP introduced a new handler called `onBackgroundEvent` which is called when a scheduled background event occurs.

> It has one parameter - `request`, that is provided without change.

Reviewer 2: What do you mean by this?

Author: Added a comment

> This SIP introduced a new handler called `onBackgroundEvent` which is called when a scheduled background event occurs.

Reviewer 2: ```suggestion
This SIP introduces a new handler called `onBackgroundEvent` which is called when a scheduled background event occurs.
```
